### ISSUE TITLE
Different anchors for Stamps

### ISSUE THREAD

**Issue by eishiya** [mapeditor/tiled#3250]
Multi-tile Stamps are currently always placed centered on the cursor. This can be a problem when trying to place stamps with variants of differing sizes precisely. For example, if I want to draw a line of trees, the shorter trees will be placed higher than the taller trees.

For stamps like this, it would be very helpful to be able to set an anchor point, so that I could choose e.g. bottom center for trees.

Since this is only relevant when using saved stamps with variants, perhaps anchors could be saved per stamp/variant, using a dropdown like Tilesets' Object Alignment, or a graphical selector:
![image](https://user-images.githubusercontent.com/70545646/150618710-d12ff216-623e-4f34-9a57-a71809438840.png)
At a minimum, I think the options should be Center, since that's the usual behaviour, and Bottom, as that's very common in both sidescrollers and top-down games, useful for any stamps that are attached to the ground. Other anchors are useful less often, but should still be offered if it's not too difficult. Saving this value on a per-stamp basis means that the user won't have to change any tool properties when they switch swamps, and won't have to think about alignment at all when it doesn't matter.

The default value for stamps should be Center to match with the current behaviour, and newly added variants should default to the value used to the previous stamp/variant within that stamp.

**Comment by CelticMinstrel** [comment]
This is a great idea, but I also partially disagree with the conclusion.

This is _not_ only relevant for saved stamps! Copy-paste is, currently, also a stamp. Being anchored always in the centre makes it _very_ difficult to paste very large chunks of tiles.

For saved stamps, I agree that there should be something to set the anchor (you could probably choose from one of 9 options from a 3x3 grid). I get that it's most important if the stamp has variations of different sizes, but I don't think it's _only_ useful in that case.

For ad-hoc stamps, I think the anchor should be automatically set based on how you built up the stamp.

If you make an ad-hoc stamp by dragging from top left to bottom right (either in the tileset palette, or on the map with right-click), then it should be anchored at the bottom right – the tile you released your mouse on. Similarly, if you drag from top right to bottom left, it should be anchored at the bottom left. This default could also be carried over into the tile stamp palette when you save a stamp, though I'm not sure how useful it is there.

I'm also unsure how to determine the anchor for stamps pasted from the clipboard… unless it's viable to include that information when copying? Make the selection remember which way you dragged to create it?

**Comment by eishiya** [comment]
Perhaps an anchor selector like in the OP could be part of the tool options for all tools that use stamps (Stamp Brush, Fill, not sure what else)? That way every stamp can benefit from anchor flexibility. No selection shenanigans would be needed, and it would provide more options than possible with selections.

And it wouldn't close the door on more specific anchors on saved tilestamps - an anchor present in a stamp would override the current tool anchor.

**Comment by CelticMinstrel** [comment]
Yeah, you could add an anchor selection to the tool options (that would allow selecting centre anchors even for ad-hoc tiles), but I would still prefer it to automatically decide the anchor based on how I dragged the mouse.

**Comment by eishiya** [comment]
If selection-based anchoring is added, I'd prefer it to be optional, as I'd prefer not to think about how I'm selecting things, whether it be from a Tileset or the map. Perhaps it can be a "from selection start" checkbox alongside the anchor selector that disables the anchor selector, or (less clearly/discoverably) the behaviour when you deselect any anchor, e.g. by clicking the same anchor you already had selected.
For example, when I select a large tree, I often start with the crown (probably from top left or top right) and then add the trunk. This would result in a top left or top right alignment, when I'd prefer bottom centre (or better yet, at the base of the trunk, which may not be the exact middle).

I think "from selection start" would be a clearer way to do it, as it removes any doubt about direction and most doubt about stamps from compound selections - the first tile/cell selected is the anchor, and the exact way you build up the rest of the selection doesn't matter. For simple rectangular selections like you've been describing, it'll still work exactly the same as if the direction was taken into account.
This would even allow fine control over the exact anchor. For example, if you want a bottom centre anchor, you could click the bottom centre tile, and then shift+select the rest of the tiles. It might get a little confusing if you then deselect the first tile, but it's not that unusual to want an empty tile to be the anchor of a complex stamp.
This way, because there's a well-defined start/anchor to the selected region, this can potentially be saved alongside the stamp when the stamp is saved or copy+pasted, which would mean that saved stamps could have arbitrary anchors assigned without any manual editing of the saved stamps (see #3273).

**Comment by CelticMinstrel** [comment]
> If selection-based anchoring is added, I'd prefer it to be optional, as I'd prefer not to think about how I'm selecting things, whether it be from a Tileset or the map. Perhaps it can be a "from selection start" checkbox alongside the anchor selector that disables the anchor selector, or (less clearly/discoverably) the behaviour when you deselect any anchor, e.g. by clicking the same anchor you already had selected.

I think it could probably just be an option separate from the 9 "specific anchor" options. You can either select the anchor you want (eg top left) or choose "auto" which bases it on the mouse movement.

Minor disadvantage of that is that you can't see in the toolbar which anchor was automatically selected… but you can see that from the paint preview, so that doesn't seem too bad.

Selection-based anchoring might be bad for non-rectangular selections. But rectangular selections are pretty common, at least for me. (Maybe because I'm used to RPGMaker where all selections are rectangular.)

> I think "from selection start" would be a clearer way to do it

No. The reason it needs to be "from selection end" is that the start of the selection might not be in view anymore. If you right-drag a huge block of tiles, you should just be able to shift the mouse a few tiles right and paste it in without altering the block's vertical positioning. If the anchor is at selection start, the entire block would be shifted down by its height when doing that.

Your description is probably fine for smaller selections where the entire selection fits in the viewport, but it's terrible for gigantic selections. It's true that my description makes no sense for discontiguous selections though.

Maybe a compromise could be that the anchor is the first tile the mouse is released on when selecting. If you shift-click to select tiles one at a time, that boils down to "selection start". If you drag a rectangle of tiles, it boils down to "selection end". If you drag a rectangle of tiles and then shift-click a few extras in, the anchor still ends up wherever you ended the rectangle, but that seems like a bit more of a niche case? Dragging a rectangle and shift-clicking some unwanted ones out seems more likely to me (is that possible?).

Side note: the 3x3 visual selector you showed as a sample in the opening post might actually be nice to use for the object aligment selector as well.
